contracts/bridge.sol

```diff
@@ -99,63 +107,117 @@ contract HomeBridge {
     }
 
     /// Constructor.
-    function HomeBridge (uint n, address[] a) public {
-        require(n != 0);
-        require(n <= a.length);
-        requiredSignatures = n;
-        authorities = a;
+    function HomeBridge (
+        uint requiredSignaturesParam,
+        address[] authoritiesParam,
+        uint estimatedGasCostOfWithdrawParam
+    ) public
+    {
+        require(requiredSignaturesParam != 0);
+        require(requiredSignaturesParam <= authoritiesParam.length);
+        requiredSignatures = requiredSignaturesParam;
+        authorities = authoritiesParam;
+        estimatedGasCostOfWithdraw = estimatedGasCostOfWithdrawParam;
     }
 
     /// Should be used to deposit money.
     function () public payable {
         Deposit(msg.sender, msg.value);
     }
 
+    // layout of message :: bytes:
+    // offset  0: 32 bytes :: uint (little endian) - message length
+    // offset 32: 20 bytes :: address - recipient address
+    // offset 52: 32 bytes :: uint (little endian) - value
+    // offset 84: 32 bytes :: bytes32 - transaction hash
+
+    // bytes 1 to 32 are 0 because message length is stored as little endian.
+    // mload always reads 32 bytes.
+    // so we can and have to start reading recipient at offset 20 instead of 32.
+    // if we were to read at 32 the address would contain part of value and be corrupted.
+    // when reading from offset 20 mload will read 12 zero bytes followed
+    // by the 20 recipient address bytes and correctly convert it into an address.
+    // this saves some storage/gas over the alternative solution
+    // which is padding address to 32 bytes and reading recipient at offset 32.
+    // for more details see discussion in:
+    // https://github.com/paritytech/parity-bridge/issues/61
+
+    function getRecipientFromMessage(bytes message) public pure returns (address) {
+        address recipient;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            recipient := mload(add(message, 20))
+        }
+        return recipient;
+    }
+
+    function getValueFromMessage(bytes message) public pure returns (uint) {
+        uint value;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            value := mload(add(message, 52))
+        }
+        return value;
+    }
+
+    function getTransactionHashFromMessage(bytes message) public pure returns (bytes32) {
+        bytes32 hash;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            hash := mload(add(message, 84))
+        }
+        return hash;
+    }
+
+    /// to be called by authorities to check
+    /// whether they withdraw message should be relayed or whether it
+    /// is too low to cover the cost of calling withdraw and can be ignored
+    function isMessageValueSufficientToCoverRelay(bytes message) public view returns (bool) {
```

Comment:
what is the difference between `const`,`pure` and `view`?

Reply:
`constant` = `view` = does not modify state

`pure` = does not read or modify state

explained in more detail here:
https://solidity.readthedocs.io/en/develop/contracts.html#functions

Reply:
`pure` = pure in the functional programming sense. same args -> same return every time